requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+2. When after signing up `posthog.identify('userid')` is called, we associate the same person_id `p1` with the new distinct id `userid`
+3. If the user later visits the site with a new device, a new anonymous distinct id `anonid2` is created. Posthog internally associates these events with a new person id `p2` as no mapping for that id exists.
+4. If after signing in `posthos.identify('userid')` is called, we associate events going forward with person_id `p1`
+
+As a result, events done between steps 1, 2 and 4 will be tracked as from person `p1` but events between steps 3 and 4 will be under person`p2`. As a result no events are lost but some will be considered as coming from a different (anonymous) person. In the above image this corresponds to the green period.
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
+
+#### 2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+Example: Count of active users broken down by `paying` user property each month in 2021.
+
+After this change, if `paying` property of a user in Dec 2021 is changed to `false`, count of `paying` users does not change for e.g. March 2021. The  user will be considered non-paying until their `paying` property changes again2.
+
+It also means users could be counted as both `paying=true` and `paying=false` during a given month when they change state.
+
+User properties will change over time on events, some customers may rely on the existing behavior (where the latest property is assumed for all prior events), however the new behavior is considered much more accurate by our customer feedback.
+
+Our proposed mitigation or most use-cases (e.g. events for users who have paid at some point) cohorts can be used to achieve the same results, we will document this for affected users
+
+
+### Technical implications
+
+### Benefits
+1. Faster queries: remove the need to do JOINs for any insight queries.]
+2. Unlock: Our query results no longer change unless new data has come in - we make it possible to later implement time-series caching to speed up repeated queries over long time frames (>10x improvement expected on repeated calculations).
+4. We can sample by person_id in queries, which will make it possible to build a quick way to speed up queries in huge (1Bn+ events) installations with no additional query time resource cost.
+5. Queries will require less memory and no longer OOM (making it cheaper and more reliable to operate posthog)
```

Comment:
This benefit refers back to the opening section - currently queries from clients with 10M+ persons will OOM due since the whole persons dataset needs to be loaded into memory for joins.

By removing the JOIN we also reduce memory usage significantly and there's no more conceptual roadblock. Yes, badly implemented queries _might_ still OOM but that's out of scope for this proposal.

Reply:
> Yes, badly implemented queries might still OOM

badly set resource requirements too on self hosted

> but that's out of scope for this proposal.

Agreed. 

I seem to have worded my point badly. I was wondering if there's a bigger benefit here of "memory usage more predictable/less spiky" which to me is a bigger selling point as that would make self hosted users and our life easier for managing their instance compared to "no longer OOM for specific customers on specific instances"

Reply:
I think we're largely on the same page here - potato potata.

Reason I prefer current wording is that the problem users are facing right now is that some of their queries will fail due to running out of memory on a single individual query due to needing to JOIN `persons` table. Clickhouse memory usage for queries is inherently spiky due to people running queries very rarely from a systems standpoint.